doc/user/guide/evaluation.rst

```diff
@@ -1,2 +1,172 @@
 Interpreting and Evaluating Expressions
 =======================================
+
+Evaluation Trace Analysis Using Nushell
+---------------------------------------
+
+`Nushell <https://www.nushell.sh>`__ can be installed from ``Homebrew``.
+
+::
+
+   brew install nushell
+
+Save this file as ``sequant.nu`` in the same directory where the trace
+output file is present. We will call this directory 'working directory'.
+
+.. code:: nu
+
+
+   use std/assert
+
+   const pattern = {
+     trace: '^\s*(Term|Eval|Cache)\s+'
+     separator: '|'
+   }
+
+   def "read log" [file: string] {
+       grep -n -E -e $pattern.trace $file
+         | parse --regex '(?<line>\d+):\s+(?<trace>.*)'
+         | each {
+           let line = $in.line
+           let fields = $in.trace
+                        | split row $pattern.separator 
+                        | str trim
+           match $fields {
+             [$tag0 $tag1 ..$rest] => {
+                         {tag0 : $tag0
+                          tag1 : $tag1
+                          trace: $rest
+                          line : ($line | into int) } }
+             }
+           }
+   }
+
+   def "new-cols eval" [] {
+       {
+         time   : ($in.0 | into duration)
+         memory : ($in.1 | into filesize)
+         annot  : $in.2
+       }
+   }
+
+   def "new-cols cache" [] {
+     {
+       key    : $in.0
+       life   : $in.1
+       nalive : ($in.2 | into int)
+       memory : ($in.3 | into filesize )
+     }
+   }
+
+   def make-tag [tag: string, make_col: closure] {
+     let entry = $in | where tag0 == $tag
+     let new_cols = $entry.trace | each { do $make_col }
+     $entry | select tag0 tag1
+            | merge $new_cols
+            | merge ($entry | select line)
+   }
+
+   export def "trace cache" [file: string] {
+     read log $file | make-tag 'Cache' { new-cols cache }
+   }
+
+   export def "trace eval" [
+               file: string # The text file with sequant evaluation trace.
+               ] {
+     read log $file | make-tag 'Eval' { new-cols eval }
+   }
+
+   export def "trace term" [file: string] {
+     let lines = read log $file | where tag0 == 'Term'
+     let begs  = $lines | every 2
+     let ends  = $lines | skip 1 | every 2
+     $begs | zip $ends
+           | each { match $in { [$b, $e] => { assert ($b.trace.0 == $e.trace.0)
+                                              { tag0  : $b.tag0
+                                                begin : $b.line
+                                                end   : $e.line
+                                                expr  : $b.trace.0
+                                              } } } }
+   }
+
+This nu script exports the functions: ``trace eval``, ``trace cache``,
```

Comment:
this script needs to be in SQ source somewhere, rather than in documentation. Docs would be a lot cleaner, and it could be tested.

Reply:
Added `doc/user/guide/sequant.nu`